LIP-37: EDF (#121)
lidofinance/lido-improvement-proposals: LIPS/lip-37.md

---
lip: 37
title: "Key Delegation Authority"

> status: WIP

Reviewer: Change to Proposed before merge

author: Raman Siamionau, Matsvei Talstalutski

> discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>

Reviewer: To be fixed before merging

### A standard primitive for permissioned key handling

Today each role handles its signing keys in its own one-off way, and rotating any of them runs through the same heavyweight governance path. Without a shared primitive, every new role re-invents key custody and rotation, and monitoring must learn each role's own conventions — multiplying operational risk. Standardizing delegation as a single, audited mechanism lets any role inherit the same rotation, revocation, and termination semantics, and lets monitoring treat every delegated key uniformly.

KDA improvements unlock:

- **Routine rotation as hygiene.** Periodic hot-key rotation limits the value of any single leaked key and shrinks the window in which an undetected compromise stays useful — a standard operational practice that the governance-gated path makes impractical today.
- **Smart-account owners.** Owners increasingly prefer a smart account (multisig or account-abstraction wallet) over a bare EOA for stronger custody, but today's permission model assumes an EOA signer. Verifying the delegatee through ERC-1271 lets the hot key itself be a contract.

> - **Future signing-scheme migration.** The signing schemes these keys rely on are not fixed forever — most concretely, an eventual migration to post-quantum signatures. Routing all signature verification through KDF means such a migration needs no per-role protocol change.

Reviewer: This one is not fully clear. In case of some special PQ scheme we might need to have new KDA, but this is not guaranteed to be needed

Author: Adjusted

KDF removes the governance from rotation. A suspect key is de-authorized immediately and a replacement comes online after a short safety cooldown, turning both routine rotation and incident response into a local operator action measured in minutes rather than the ~10 days a governance vote takes today.

## Specification

### Overview

The release consists of three coordinated changes:

1. Deployment of the `DelegationFactory`, establishing the general-purpose on-chain delegation infrastructure (KDF).
2. An update to the `DepositSecurityModule` contract to verify guardian signatures via the guardian's ERC-1271 `DelegationContract.isValidSignature`, so that Council guardians can operate behind KDF delegation contracts.

> 3. Modifications to the Oracle and Council daemons to support operating behind delegation contracts.

Reviewer: ```suggestion
3. Modifications to the Oracle's and Council daemons' off-chain code to support operating behind delegation contracts.
```

Author: Fixed

**Why a purpose-built own delegation contract rather than reusing an existing solution.** Other delegation frameworks we looked into are far more complex and flexible than this role needs, and that unused flexibility is attack and misconfiguration surface. A minimal, non-upgradeable contract instead makes KDF's guarantees structural — owner can never `execute()` or sign, one active delegatee, immediate revocation with a cooldown-gated assignment, irreversible `terminate()` — keeping the audited surface small.

**Why keep cryptography outside the protocol.** All signature verification lives in KDF, so the eventual move to post-quantum signatures means updating only KDF, not the protocol. The protocol just does a generic ERC-1271 check against the delegation contract address and never needs to know how a signature is produced, so a PQ migration stays entirely within KDF.

### Technical Specification

---

#### Part 1: Key Delegation Framework (KDF) — Contracts

##### Architecture

> The delegation layer is composed of two contracts deployed from a single factory:

Reviewer: from a single factory or repo? Because I see that one of the contracts is factory itself

Author: fixed

- **`DelegationFactory`** — a singleton that deploys a fresh `DelegationContract` per instance. The per-instance `owner` and `cooldown` are constructor arguments stored as Solidity `immutable`s; the initial `delegatee` is set in the same constructor.
- **`DelegationContract`** — a non-upgradeable, minimal on-chain delegation contract. It enforces a one-owner / one-active-delegatee model and supports two integration patterns described below.

##### Roles

The contract has exactly two trusted entities, **owner** and **delegatee**:

| Role          | Custody                         | Capabilities                                                                                                                                      |
|---------------|---------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|

> | **Owner**     | Safe multisig or cold wallet    | Assign, reassign, and revoke the delegatee (`assignDelegatee()` / `revokeDelegatee()`); irreversibly `terminate()` the contract                   |

Reviewer: ```suggestion
| **Owner**     | Safe multisig or cold wallet or AA wallet | Assign, reassign, and revoke the delegatee (`assignDelegatee()` / `revokeDelegatee()`); irreversibly `terminate()` the contract                   |
```

Author: I'm not sure we should recommend or propose using AA here. IMHO, there should ultimately be a cold wallet or a multisig acting as the owner.
AA allows transfering ownership, which doesn't align with our preference to handle such changes through governance.

    /// @param owner     The contract's owner, set as a constructor immutable.
    ///                  Fixed for the lifetime of the contract; replacing 
    ///                  the owner requires deploying a new contract.
    /// @param delegatee Initial active delegatee, set in the constructor (and
    ///                  effective immediately — the cooldown applies only to
    ///                  later reassignments). Mutable thereafter via
    ///                  assignDelegatee(). Pass address(0) to deploy with no
    ///                  delegatee.
    /// @param cooldown  Seconds a reassigned delegatee waits before becoming
    ///                  effective (see assignDelegatee). Set as a constructor
    ///                  immutable; may be 0 to disable the cooldown.
    /// @return instance Address of the newly deployed DelegationContract.
    function deploy(address owner, address delegatee, uint256 cooldown)
        external
        returns (address instance);

    /// @notice Emitted for each DelegationContract deployed by the factory.
    event DelegationContractDeployed(
        address indexed instance,
        address indexed owner,

>         address delegatee,

Reviewer: Can also be indexed since you have 1 indexed topic left and one might search for the contracts where the initial delegate was a particular address

Author: I've decided that's it quite useless, since delegatee can be changed.
Added indexed back

        uint256 cooldown
    );
}

interface IDelegationContract {
    // --- Owner controls ---

    /// @notice Assign (or reassign) the active delegatee.
    ///         Only callable by owner. The new delegatee becomes effective only
    ///         after the contract's cooldown (`getCooldown()` seconds, or
    ///         immediately if cooldown is 0). The currently effective delegatee
    ///         (if any) stays effective throughout the cooldown and is dropped
    ///         only when the new one activates — so a planned rotation is
    ///         seamless, with no interval lacking an active delegatee.
    ///         Reassigning before the cooldown elapses replaces the pending
    ///         delegatee and restarts the cooldown; the current one stays
    ///         effective throughout. To drop a (e.g. compromised) delegatee
    ///         immediately, use revokeDelegatee().
    ///         Reverts if delegatee == owner.
    /// @param delegatee Address of the incoming delegatee.
    function assignDelegatee(address delegatee) external;

    /// @notice Remove the current and pending delegatee. 
    ///         Only callable by owner.

>     function revokeDelegatee() external;

Reviewer: Do these two methods also revert when the contract is terminated? It is mentioned below for execute, but not mentioned here

Author: Added

    function execute(address target, bytes calldata data)
        external
        payable
        returns (bytes memory result);

    // --- Pull integration (ERC-1271) ---

    /// @notice ERC-1271 signature validation. Returns the ERC-1271 magic value
    ///         (0x1626ba7e) if `signature` is a valid signature over `hash`
    ///         by the contract's current *effective* delegatee; otherwise
    ///         returns 0xffffffff.
    ///         The delegatee is resolved via getDelegatee(), so validation
    ///         fails closed when there is no effective delegatee (never
    ///         assigned, revoked, or terminated → address(0)).
    ///
    ///         NOTE: unlike a raw ECDSA check, this result is state-dependent
    ///         and revocable — it can return valid at one block and invalid at
    ///         the next (e.g. after the delegatee is rotated, revoked, or the
    ///         contract is terminated).
    /// @param hash      Message hash that was signed.
    /// @param signature Opaque signature bytes (ECDSA).
    function isValidSignature(bytes32 hash, bytes calldata signature)
        external
        view
        returns (bytes4 magicValue);

>     // --- Interface detection (ERC-165) ---

Reviewer: What about token recovery methods? I assume that in some cases, ETH or other tokens can be directed to this address. Since the contract is immutable, it will be great to have AssetRecoverer added

    ///         timestamp at which it becomes effective, or (address(0), 0) if
    ///         no assignment is in cooldown.
    function getPendingDelegatee() external view returns (address delegatee, uint256 activeFrom);

    /// @notice Cooldown in seconds between assigning a delegatee and it
    ///         becoming effective. Set in the constructor (Solidity immutable)
    ///         and unchangeable thereafter.
    function getCooldown() external view returns (uint256);

    /// @notice Returns true if the contract has been terminated.
    function isTerminated() external view returns (bool);

    // --- Events ---

    event DelegateeNominated(address indexed newDelegatee, uint256 activeFrom);
    event DelegateeRevoked(address indexed revokedDelegatee);
    event Terminated();
}
```

##### Integration Models

The `DelegationContract` is designed as a **general integration primitive** for any permissioned bot in the Lido ecosystem. Two complementary patterns are supported; protocol integrations should select the one that fits their verification model.

###### Pull Style — ERC-1271 Signature Verification

> The `DelegationContract` is itself an **[ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) signer**: it implements `isValidSignature(hash, signature)`. The integrator (protocol) contract treats the `DelegationContract` address as the signing principal and verifies a relayed signature via OpenZeppelin's [`SignatureChecker.isValidSignatureNow(delegationContract, hash, signature)`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/SignatureChecker.sol).

Reviewer: I'm not sure if we need `isValidSignatureNow` on the protocol level. I suggest that on the protocol level we use pure ERC-1271 check, and use OZ's `isValidSignatureNow` within the delegation contract itself. Hence, there will be no way to use EOA with the protocol

Author: fixed

All of the delegation indirection lives inside `isValidSignature`: it resolves the current effective delegatee via `getDelegatee()` and validates the signature against it. This keeps the integrator **delegation-agnostic**. The contract fails closed: with no effective delegatee (never assigned, revoked, or terminated) — `getDelegatee()` is `address(0)`.

The integrator must know which `DelegationContract` to check. The address must be relayed alongside the signature, so the signature is bound to a specific delegation contract and the integrator reads the target from the message.

> The contract also implements [ERC-165](https://eips.ethereum.org/EIPS/eip-165) `supportsInterface`, advertising ERC-1271, [ERC-5313](https://eips.ethereum.org/EIPS/eip-5313), and `IDelegationContract`. The owner is exposed via the ERC-5313 `owner()` view so block explorers and multisig tooling recognize the controlling party; there is deliberately no `transferOwnership`, preserving owner immutability.

Reviewer: This block does not seem to belong to this section. This is not about pull style but a general consideration

Author: removed

```
Off-chain bot (delegatee)          DelegationContract              Protocol contract
        │                                  │                              │
        │  sign message (hot key)          │                              │
        │─ signature + DelegationContract addr relayed off-chain ────────►│
        │                                  │                              │
        │                                  │◄ isValidSignature(hash,sig) ─│
        │                                  │                              │
        │                                  │─ 0x1626ba7e (or 0xffffffff) ►│
        │                                  │                              │
        │                                  │   integrator also checks the │
        │                                  │   addr is a registered       │
        │                                  │   principal (e.g. guardian)  │
```

**Security assumptions:**

> - The integrator must treat the `DelegationContract` address — not the signing key — as the authorized principal: it checks that the supplied `DelegationContract` is a registered entity **and** that `isValidSignature` returns the magic value.

Reviewer: What's 'magic value'?

Author: Added more details

- The signing key must match the effective delegatee at the time of on-chain verification. A key rotated (or the contract terminated) after signing but before verification causes `isValidSignature` to reject — failing closed.
- The fail-closed-on-zero-delegatee guard lives **inside** `isValidSignature` (it returns the failure value rather than verifying against `address(0)`), so integrators cannot accidentally validate a malformed signature against an unassigned delegatee.

Reference implementation of the contract side:

```solidity
function isValidSignature(bytes32 hash, bytes calldata signature)
    external
    view
    returns (bytes4)
{
    address delegatee = getDelegatee();   // effective delegatee, or address(0)
    // Fail closed when there is no effective delegatee (never assigned/revoked/terminated).
    if (delegatee != address(0) && SignatureChecker.isValidSignatureNow(delegatee, hash, signature)) {
        return IERC1271.isValidSignature.selector;   // 0x1626ba7e
    }

>     return 0xffffffff;

Reviewer: Is there a constant or smth similar to `IERC1271.isValidSignature.selector` for the failure scenarios?

Author: No, there is not

}
```

> **When to use:** Any integration where the protocol collects signatures off-chain and verifies them on-chain in a single transaction.

Reviewer: Why in a single transaction? What stops us from using it when signatures are checked in several TXs?

Author: removed

###### Push Style — Transaction Execution via `execute()`

The delegatee calls `execute(target, data)` on the `DelegationContract`, which forwards the call to the target protocol contract. From the target's perspective, `msg.sender` is the `DelegationContract` address.

```
Off-chain bot (delegatee)          DelegationContract         Protocol contract
        │                                  │                          │
        │── execute(target, data) ────────►│                          │
        │   [optional msg.value]           │                          │
        │                                  │── call(target, data) ───►│
        │                                  │   msg.sender ==          │
        │                                  │   DelegationContract     │
        │                                  │◄── result ───────────────│
        │◄── result (or revert) ───────────│                          │
```

> `execute()` is `payable` and forwards `msg.value` to the target. This is required to support future permissioned bots that interact with payable protocol functions where a fee must be paid in ETH at call time. After the target call returns, `execute()` sweeps any ETH still held by the contract back to the delegatee, so change refunded by the target is forwarded to the bot rather than stranded on the delegation contract.

Reviewer: I suggest we have a dedicated method for ETH recovery and return only the difference between the payment supplied and the payment spent.

Reviewer: Still relevant because of the "weeps any ETH still held by the contract back to the delegatee". We should return only the diff between the payment amount provided and the actual payment amount used. Otherwise, any ETH that accidentally appeared on the contract will be credited to delegatee

Reviewer 2: Well, agreed. 

Added recoverETH method, which can be used by owner only.

Check out commit: https://github.com/lidofinance/lido-improvement-proposals/pull/121/changes/8a386b17d12aa34a26d1e714e9e010fb4cf52c03

- **Revocation** (`revokeDelegatee()`) is immediate and clears both the current and any pending delegatee, leaving the contract with no delegatee until a new one is assigned and its cooldown elapses.

The cooldown is set in the constructor and cannot change; it may be 0.

##### Emergency Termination

The owner may call `terminate()` to permanently disable the contract's `execute()` function; it also clears the active delegatee (equivalent to `revokeDelegatee()`) as part of the same call. This is an **irreversible** operation: a terminated contract cannot be reactivated. After termination:

> - All delegatee calls to `execute()` revert.

Reviewer: I suggest we also lock the `assignDelegate` method

Author: fixed

- `getDelegatee()` returns `address(0)`, so any pull-style integrator that resolves the active delegatee through it (to verify a signature) fails closed and does not keep trusting the last delegatee.
- A new `DelegationContract` must be deployed and the role must be reassigned to the new address.

Termination is intended for scenarios where the owner believes the owner's own address has been compromised and wishes to ensure no further actions can be taken under its authority.

> ##### Hot-Key Rotation

Reviewer: We need a note on how on-chain monitoring from the Lido side will be performed. If the factory is permissionless, we will need to automatically fetch designated role members for the pre-defined roles that are expected to be KDF contracts. For these contracts, we need to have alerts about delegate changes and termination. The list of KDF contracts should not be static to prevent bling spot after rotations.

Author: Let's not add this to the LIP. I think we can align on this internally in an internal document.

---

#### Part 2: KDF Adoption — Oracle and DSM Integration

This section defines which components are migrated to the delegation model in this release, the concrete changes required for each, and the operator migration process. Adoption is mandatory for Oracle and DSM roles.

##### Scope of Migration in This Release

| Component                              | Integration style                                                                                                            |
|----------------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| Lido Oracle                            | Push (`execute()`)                                                                                                           |
| `DepositSecurityModule` contract       | Pull (DSM verifies via a standard ERC-1271 `isValidSignature` check against the guardian's `DelegationContract`)             |
| Council daemon                         | Pull (signs messages, publishes its `DelegationContract` address with each signature, DSM verifies)                          |

> | Depositor bot                          | Claims the council-provided `(signature, DelegationContract address)` pairs; re-sorts the deposit array by guardian address. |

Reviewer: This line is unclear to me

Author: Depositor should support the new type of messages and sort them correctly using the guardian address instead of the sender address.

**Pattern:** Push via `execute()`.

**Authorization model:** `HashConsensus` and the report processors (`AccountingOracle`, `ValidatorsExitBusOracle`, CM's `FeeOracle`, and CSM's `FeeOracle`) check `msg.sender` against the registered oracle committee member set. After migration, `msg.sender` is the `DelegationContract` address, which governance must have registered as the committee member.

**Daemon configuration:** `DELEGATION_CONTRACT_ADDRESS` environment variable. When set, all oracle contract calls are routed through `DelegationModule`, a Web3py extension that wraps `execute()` dispatch. Startup validation checks: (a) `getDelegatee() == configuredHotKey`, and (b) `delegationContract` address holds oracle committee membership in `HashConsensus`.

**Transitional behavior:** When `DELEGATION_CONTRACT_ADDRESS` is unset, calls are sent directly from the hot key EOA as before. This path exists for development, for the migration window, and as a long-term fallback.

---

###### DSM Contract — Pull Integration

**Pattern:** Pull — DSM verifies each signature against the supplied guardian with a standard ERC-1271 `isValidSignature` check over a digest **bound to that guardian**.

> **Authorization model today.** `DepositSecurityModule` does **not** take the guardian address as an input — it *derives* it: every verification path calls `ECDSA.recover(msgHash, sig.r, sig.vs)` and looks the recovered EOA up in the guardian set via `_isGuardian`.

Reviewer: I thought we planned to remove `ECDSA.recover(msgHash, sig.r, sig.vs)` entirely. We need to come up with a different approach here

Author: yeah, it's todays authorization model we want to change

function _effectiveSigner(address guardian) internal view returns (address) {
    return IDelegationContract(guardian).getDelegatee();
}

address prevGuardian;
address[] memory signers = new address[](sortedGuardianSignatures.length);
for (uint256 i = 0; i < sortedGuardianSignatures.length; ++i) {
    GuardianSignature calldata gs = sortedGuardianSignatures[i];

    bytes32 msgHash = _hashDepositMessage(gs.guardian, /* ...existing fields... */);
    // _isValidGuardianSignature checks _isGuardian + the guardian's ERC-1271 signature.
    if (!_isValidGuardianSignature(gs.guardian, msgHash, gs.signature)) revert InvalidSignature();

    // Distinct guardians: strictly ascending by guardian address.
    if (gs.guardian <= prevGuardian) revert GuardiansNotSortedOrDuplicate();
    prevGuardian = gs.guardian;

    // Distinct hot keys: no resolved delegatee may repeat across the batch.
    // The guardian set is small (quorum is 4-of-6), so the nested scan is cheap.
    address signer = _effectiveSigner(gs.guardian);
    for (uint256 j = 0; j < i; ++j) {
        if (signers[j] == signer) revert DuplicateSigner();
    }
    signers[i] = signer;

> }

Reviewer: function wrapper is missing in the code

Author: It's part of _verifyAttestSignatures function. Added note above

```

Strictly-ascending-by-guardian rejects both an unsorted array and any duplicate guardian in a single comparison; the nested scan then rejects any repeated resolved delegatee, covering the case where two distinct guardian contracts resolve to the same hot key. Each `signer` is non-zero, because `_isValidGuardianSignature` has already rejected a guardian with no effective delegatee. The submitter must sort the batch by **guardian address** to satisfy the ascending check.

**Affected DSM entry points:** each `Signature` / `Signature[]` argument is extended to carry the guardian address as above:

```solidity
function depositBufferedEther(
    uint256 blockNumber,
    bytes32 blockHash,
    bytes32 depositRoot,
    uint256 stakingModuleId,
    uint256 nonce,
    bytes calldata depositCalldata,

>     GuardianSignature[] calldata sortedGuardianSignatures   // was Signature[]

Reviewer: Use Ethereum EIP style notation here and in all doc

```suggestion
    // Changed in LIP-37, was Signature[] before
    GuardianSignature[] calldata sortedGuardianSignatures
```

Author: fixed

**Required change to the council daemon:** Sign with the delegatee private key, set `DELEGATION_CONTRACT_ADDRESS`, and publish that `DelegationContract` address together with each signature.

**Required change to the depositor bot:** The bot relays the council-provided `(DelegationContract address, signature)` pairs to DSM as-is. Two behavioral changes are required:

1. **Filtering (on receival):** accept an incoming message only if its signer is the **current delegatee** of the referenced guardian's `DelegationContract` (resolved via `getDelegatee()`), instead of matching against a delegation-contract address. Messages signed by a rotated-out key are dropped.
2. **Ordering (on submission):** when assembling `depositBufferedEther`, sort the signature array by **guardian address**, matching DSM's strictly-ascending-by-guardian check.

---

###### Validator Ejector (Node-Operator Side) — Off-Chain Update

The **validator ejector** is the off-chain daemon run by each node operator that watches `ValidatorsExitBusOracle` (VEBO) — the oracle that publishes which validators must exit — and broadcasts the corresponding signed voluntary-exit messages to the consensus layer.

> To accept those messages the ejector keeps an **allowlist** of trusted oracle signing addresses. Oracle messages are always signed by the **delegatee EOA**, never by the `DelegationContract`, so the allowlist must contain the **delegatee EOA** and be refreshed to the new EOA whenever the delegatee is rotated.

Reviewer: Sound not ideal, but I guess there is no other way here. We need to think about future PQ or AA delegates. How will we fetch TX sender in this case?

Author: In this case, we should require the NO to host the EL node along with the Ejector and drop this verification at all.

Concretely, we propose to deploy a new `DelegationFactory` / `DelegationContract` system (repository: [`lidofinance/delegation-execution-authority`](https://github.com/lidofinance/delegation-execution-authority)), and targeted updates to the [`lido-oracle`](https://github.com/lidofinance/lido-oracle) daemon and [`DepositSecurityModule`](https://docs.lido.fi/contracts/deposit-security-module). The adoption scope specifies the integration points: the `lido-oracle` daemon gains delegation support via a `DELEGATION_CONTRACT_ADDRESS` environment variable, and the `DepositSecurityModule` is updated to verify guardian signatures with a standard ERC-1271 check against the guardian's `DelegationContract` (which implements `isValidSignature`), enabling council members to operate behind delegation contracts. Adoption is mandatory for these roles; a coordinated migration process is defined for existing Oracle and DSM operators to onboard to the new model without service interruption.

## Motivation

> Operational compromise is now the dominant cause of losses in the ecosystem. Of the roughly $4 billion lost to Web3 incidents in 2025, about $2.1 billion (~52%) stemmed from access-control and operational-security failures — leaked signing keys, compromised machines, and mishandled signers — against roughly $512 million from smart-contract vulnerabilities ([Hacken 2025 Security Report](https://hacken.io/insights/2025-security-report/)). The largest and least-recoverable losses now come from compromised signers, not flawed code.

Reviewer 3: in 2026 H1 it got closer to 80% though iirc

Reviewer 3: can use more recent sources jic: https://quantstamp.com/blog/june-security-beat

Reviewer: Added

```solidity
interface IDelegationFactory {
    /// @notice Deploy a new DelegationContract.
    /// @param owner     The contract's owner, set as a constructor immutable.
    ///                  Fixed for the lifetime of the contract; replacing 
    ///                  the owner requires deploying a new contract.
    /// @param delegate  Initial active delegate, set in the constructor (and
    ///                  effective immediately — the cooldown applies only to
    ///                  later reassignments). Mutable thereafter via
    ///                  assignDelegate(). Pass address(0) to deploy with no
    ///                  delegate.
    /// @param cooldown  Seconds a reassigned delegate waits before becoming
    ///                  effective (see assignDelegate). Set as a constructor
    ///                  immutable; may be 0 to disable the cooldown.
    /// @return instance Address of the newly deployed DelegationContract.

>     function deploy(address owner, address delegate, uint256 cooldown)

Reviewer 3: not sure that cooldown is something that should be controlled by the deployer 🤔

Author: Why? I think it will be useful to make this contract permissionless, so anyone can use it for their own purposes.
Before the gov vote, we will verify the deployment, the contract, and its params to make sure they match our expectations for Oracles or Councils.

        address indexed owner,
        address indexed delegate,
        uint256 cooldown
    );
}

interface IDelegationContract {
    // --- Owner controls ---

    /// @notice Assign (or reassign) the active delegate.
    ///         Only callable by owner. The new delegate becomes effective only
    ///         after the contract's cooldown (`getCooldown()` seconds, or
    ///         immediately if cooldown is 0). The currently effective delegate
    ///         (if any) stays effective throughout the cooldown and is dropped
    ///         only when the new one activates — so a planned rotation is
    ///         seamless, with no interval lacking an active delegate.
    ///         Reassigning before the cooldown elapses replaces the pending
    ///         delegate and restarts the cooldown; the current one stays
    ///         effective throughout. To drop a (e.g. compromised) delegate
    ///         immediately, use revokeDelegate().
    ///         Reverts if delegate == owner.
    ///         Reverts if the contract is terminated.
    /// @param delegate Address of the incoming delegate.
    function assignDelegate(address delegate) external;

>     /// @notice Remove the current and pending delegate. 

Reviewer 3: ```suggestion
    /// @notice Immediately remove the current and pending delegate. 
```

Author: Added

2. **Ordering (on submission):** when assembling `depositBufferedEther`, sort the signature array by **guardian address**, matching DSM's strictly-ascending-by-guardian check.

---

###### Validator Ejector (Node-Operator Side) — Off-Chain Update

The **validator ejector** is the off-chain daemon run by each node operator that watches `ValidatorsExitBusOracle` (VEBO) — the oracle that publishes which validators must exit — and broadcasts the corresponding signed voluntary-exit messages to the consensus layer.

To accept those messages the ejector keeps an **allowlist** of trusted oracle signing addresses. Oracle messages are always signed by the **delegate EOA**, never by the `DelegationContract`, so the allowlist must contain the **delegate EOA** and be refreshed to the new EOA whenever the delegate is rotated.

---

##### Operator Migration Process

Migration is designed to be zero-downtime. All preparatory steps are completed off the critical path, leaving a single irreversible governance vote that flips every prepared seat to delegation at once.

> 1. **Deploy and configure contracts (operators)**: each operator deploys its `DelegationContract` via `DelegationFactory.deploy(ownerAddress, hotKey, cooldown)` — the owner being a Safe multisig, fixed for the contract's lifetime (see Owner Immutability in Part 1); the active delegate (its existing hot key) set in the same transaction; and a `cooldown` of **1 hour** for both Oracle and DSM contracts (reduced or 0 on testnet). (Passing `address(0)` for the delegate and assigning it later via `assignDelegate()` is also supported.)

Reviewer 3: idk where 1 hour comes from 🤔

Author: fixed

---
lip: 37
title: "Execution Delegation Framework"
status: Proposed
author: Raman Siamionau, Matsvei Talstalutski

> discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
> created: 2026-06-01
> ---

Reviewer 3: ```suggestion
created: 2026-06-01
updated: 2026-07-17
---
```

---
lip: 37
title: "Execution Delegation Framework"
status: Proposed
author: Raman Siamionau, Matsvei Talstalutski
discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
created: 2026-06-01
updated: 2026-07-17
---

## Simple Summary

> This proposal introduces an Execution Delegation Framework (EDF) and adopts it across Lido's Oracle and DSM stacks:

Reviewer: ```suggestion
This proposal introduces an Execution Delegation Framework (EDF) and adopts it across Lido's Oracle ([AO](https://docs.lido.fi/contracts/accounting-oracle) and [VEBO](https://docs.lido.fi/contracts/validators-exit-bus-oracle)) and [DSM](https://docs.lido.fi/contracts/deposit-security-module) stacks:
```

1. **Execution Delegation Framework (EDF)** — a general-purpose delegation mechanism for permissioned roles, intended as protocol-recommended infrastructure for off-chain components and broader usage. The protocol grants a permission once to a delegation contract; that contract's owner then chooses and rotates the active hot signing key freely, without a governance vote — a compromised key can be de-authorized by its owner immediately, and a replacement becomes effective after a safety cooldown.

> 2. **EDF adoption within Oracle and DSM scope** — specifies the concrete off-chain and on-chain changes required to adopt EDF across [Lido Oracle](https://docs.lido.fi/holders/lido-oracle) and [Lido Council Daemon](https://docs.lido.fi/holders/lido-council-daemon).

Reviewer: ```suggestion
2. **EDF adoption within Oracle and DSM scope** — specifies the concrete off-chain and on-chain changes required to adopt EDF across [Lido Oracle](https://docs.lido.fi/holders/lido-oracle) and [Lido Council Daemon](https://docs.lido.fi/holders/lido-council-daemon) members.
```

Concretely, we propose to deploy a new `DelegationFactory` / `DelegationContract` system (repository: [`lidofinance/delegation-execution-authority`](https://github.com/lidofinance/delegation-execution-authority)), and targeted updates to the [`lido-oracle`](https://github.com/lidofinance/lido-oracle) daemon and [`DepositSecurityModule`](https://docs.lido.fi/contracts/deposit-security-module). The adoption scope specifies the integration points: the `lido-oracle` daemon gains delegation support via a `DELEGATION_CONTRACT_ADDRESS` environment variable, and the `DepositSecurityModule` is updated to verify guardian signatures with a standard ERC-1271 check against the guardian's `DelegationContract` (which implements `isValidSignature`), enabling council members to operate behind delegation contracts. Adoption is mandatory for these roles; a coordinated migration process is defined for existing Oracle and DSM operators to onboard to the new model without service interruption.

## Motivation

> Operational compromise is now the dominant cause of losses in the ecosystem. Of the roughly $4 billion lost to Web3 incidents in 2025, about $2.1 billion (~52%) stemmed from access-control and operational-security failures — leaked signing keys, compromised machines, and mishandled signers — against roughly $512 million from smart-contract vulnerabilities ([Hacken 2025 Security Report](https://hacken.io/insights/2025-security-report/)). The largest and least-recoverable losses now come from compromised signers, not flawed code.

Reviewer: ```suggestion
Operational compromise is now the dominant cause of losses in the ecosystem. Of the roughly $4 billion lost to Web3 incidents in 2025, about $2.1 billion (~52%) stemmed from access-control and operational-security failures — leaked signing keys, compromised machines, and mishandled signers — against roughly $512 million from smart-contract vulnerabilities ([Hacken 2025 Security Report](https://hacken.io/insights/2025-security-report/)). The largest and least-recoverable losses now come from compromised signers, not flawed code, and the situation is getting [even more noticeable in 2026](https://quantstamp.com/blog/june-security-beat). 
```

- Require a full on-chain governance vote (~10 days) to rotate, whether after a compromise or as a routine precaution.

By their very nature, these bots must operate with hot keys: the signing key lives on the machine, which makes it inherently exposed and comparatively easy to compromise. The liability is operational: rotating a hot key today requires a full on-chain governance vote, drawing in the dev team and token holders for what should be routine key management. That cost cuts both ways. It makes proactive, periodic rotation too expensive to do regularly, so keys live longer than they should. And when a key is actually suspected, the replacement is only authorized once the vote executes — leaving the operator to choose between continuing to run on a key it no longer trusts or dropping the seat entirely until governance acts.

EDF removes the governance from rotation. A suspect key is de-authorized immediately and a replacement comes online after a short safety cooldown, turning both routine rotation and incident response into a local operator action measured in minutes rather than the ~10 days a governance vote takes today.

## Specification

### Overview

> The release consists of three coordinated changes:

Reviewer: ```suggestion
The release consists of four coordinated changes:
```

1. Deployment of the `DelegationFactory`, establishing the general-purpose on-chain delegation infrastructure (EDF).
2. An update to the `DepositSecurityModule` contract to verify guardian signatures via the guardian's ERC-1271 `DelegationContract.isValidSignature`, so that Council guardians can operate behind EDF delegation contracts.

> 3. Modifications to the Oracle and Council daemons' off-chain code to support operating behind delegation contracts.

Reviewer: ```suggestion
3. Modifications to the Oracle and Council daemons' off-chain code to support operating behind delegation contracts.
4. Rotation of all members of Lido Oracle and Lido Council Daemon members to a deployed delegation contract instances.
```

---
lip: 37

> title: "Key Delegation Authority"

Reviewer 4: Still confused about what KDA stands for and what KDF stands for

Author: KDF (Key Delegation Framework) - is a basic contract that provides delegation functionality with recommendations on how to use it.
KDA (Key Delegation Authority) - is the name of the proposal that introduces the KDF and adopts it across Lido.

Reviewer 3: I agree with George, we use KDA and KDF together a bit inconsistently, especially the KDA term

Author: эх, changed

## Motivation

Operational compromise is now the dominant cause of losses in the ecosystem. Of the roughly $4 billion lost to Web3 incidents in 2025, about $2.1 billion (~52%) stemmed from access-control and operational-security failures — leaked signing keys, compromised machines, and mishandled signers — against roughly $512 million from smart-contract vulnerabilities ([Hacken 2025 Security Report](https://hacken.io/insights/2025-security-report/)). The largest and least-recoverable losses now come from compromised signers, not flawed code.

### A standard primitive for permissioned key handling

Today each role handles its signing keys in its own one-off way, and rotating any of them runs through the same heavyweight governance path. Without a shared primitive, every new role re-invents key custody and rotation, and monitoring must learn each role's own conventions — multiplying operational risk. Standardizing delegation as a single, audited mechanism lets any role inherit the same rotation, revocation, and termination semantics, and lets monitoring treat every delegated key uniformly.

> KDA improvements unlock:

Reviewer 4: KDF?

Author: KDA - KDF + migration and new cross-lido policy

The owner manages the delegate's lifecycle but can never act *as* the contract. Termination is the escape hatch for owner-key compromise: because `terminate()` is irreversible — it permanently disables `execute()` and clears the delegate — a still-trusted owner can neutralize the contract entirely.

##### Contract Interface

The contract implements [ERC-165](https://eips.ethereum.org/EIPS/eip-165) `supportsInterface`, advertising ERC-1271, [ERC-5313](https://eips.ethereum.org/EIPS/eip-5313), and `IDelegationContract`. The owner is exposed via the ERC-5313 `owner()` view.

```solidity
interface IDelegationFactory {
    /// @notice Deploy a new DelegationContract.
    /// @param owner     The contract's owner, set as a constructor immutable.
    ///                  Fixed for the lifetime of the contract; replacing 
    ///                  the owner requires deploying a new contract.
    /// @param delegate Initial active delegate, set in the constructor (and
    ///                  effective immediately — the cooldown applies only to
    ///                  later reassignments). Mutable thereafter via
    ///                  assignDelegate(). Pass address(0) to deploy with no

>     ///                  delegate.

Reviewer 4: ```suggestion
    /// @param delegate  Initial active delegate, set in the constructor (and
    ///                  effective immediately — the cooldown applies only to
    ///                  later reassignments). Mutable thereafter via
    ///                  assignDelegate(). Pass address(0) to deploy with no
    ///                  delegate.
```

Author: Added

    /// @notice Terminate the contract, permanently disabling execute(), signature verification via isValidSignature(), and further delegate reassignment via assignDelegate.
    ///         Only callable by owner.
    ///         Also clears the active delegate (as revokeDelegate), so
    ///         getDelegate() returns address(0) after termination.
    ///         Intended for emergency use when the owner is suspected compromised.
    ///         Termination is irreversible.
    ///         Reverts if the contract is already terminated.
    function terminate() external;

    /// @notice Recover ETH that is not attributable to an in-flight execute()
    ///         call — e.g. force-sent (selfdestruct), residual, or a stray
    ///         transfer. execute() refunds only the current call's change, so
    ///         this stray ETH is never forwarded automatically; withdrawing it
    ///         requires an explicit owner call, which may send it to any
    ///         recipient. Only callable by owner; remains callable after
    ///         termination so stranded ETH can still be rescued. Emits
    ///         ETHRecovered.
    /// @param to Recipient of the recovered ETH.

>     function recoverETH(address to) external;

Reviewer 4: I think `recoverETH` method is useless.

The contract by default doesn't accept ETH transfers and reverts, and I don't see the point in changing this behavior by defining `receive() external payable`.

Exotic methods like `selfdestruct` or specifying an address as the block's `fee recipient`, etc don't seem like something we need to handle.

Author: Done

    // --- Push integration ---

    /// @notice Execute an arbitrary non-delegate call on behalf of this contract.
    ///         Only callable by the current delegate.
    ///         Reverts if the contract is terminated.
    ///         Reverts if the target call reverts.
    ///         Forwards msg.value to the target to support payable targets.
    ///         After the target call returns, refunds only this call's change
    ///         — the unspent portion of msg.value, measured as the balance delta
    ///         over a pre-call baseline — to the delegate (msg.sender). ETH not

>     ///         attributable to this call is left untouched (see recoverETH).

Reviewer 4: The contract is not designed to receive ETH. To do that, you need `receive() external payable`. So it’s not possible to return the unspent portion of `msg.value` to it.

And I don’t think we need to add support for this contract to receive ETH. So I’d suggest revising this wording and removing the `recoverETH` method and `ETHRecovered` event.

This is a fairly specific scenario where we send a certain `msg.value` to a contract, and in response, a portion of that value may be returned specifically to `msg.sender` rather than an arbitrary address.

Note that our implementations of `TriggerableWithdrawalsGateway` and `ConsolidationGateway` allow for an arbitrary recipient, and that’s the kind of design I would expect for such cases. Otherwise, it would impose certain requirements on the sender.

Author: done

The integrator must know which `DelegationContract` to check. The address must be relayed alongside the signature, so the signature is bound to a specific delegation contract and the integrator reads the target from the message.

```
Off-chain bot (delegate)          DelegationContract              Protocol contract
        │                                  │                              │
        │  sign message (hot key)          │                              │
        │─ signature + DelegationContract addr relayed off-chain ────────►│
        │                                  │                              │
        │                                  │◄ isValidSignature(hash,sig) ─│
        │                                  │                              │
        │                                  │─ 0x1626ba7e (or 0xffffffff) ►│
        │                                  │                              │
        │                                  │   integrator also checks the │
        │                                  │   addr is a registered       │
        │                                  │   principal (e.g. guardian)  │

> ```

Reviewer 4: Consider using mermaid sequence diagrams. Smth like:

```mermaid
sequenceDiagram
    participant Bot as Off-chain bot<br/>(delegate)
    participant Anyone as Anyone<br/>(relayer)
    participant Delegation as DelegationContract
    participant Protocol as Protocol contract

    Bot->>Bot: Sign message<br/>(hot key)
    Bot->>Anyone: Send signature<br/>(off-chain)

    Anyone->>Protocol: Submit signature +<br/>DelegationContract address

    Protocol->>Protocol: Integrator checks<br/>the address is a registered<br/>principal, e.g. guardian

    Protocol->>Delegation: isValidSignature(hash, sig)
    Delegation->>Protocol: 0x1626ba7e<br/>or 0xffffffff
```

- **Revocation** (`revokeDelegate()`) is immediate and clears both the current and any pending delegate, leaving the contract with no delegate until a new one is assigned and its cooldown elapses.

> The cooldown is set in the constructor and cannot change; it may be 0.

Reviewer 4: This is more of an implementation question, but I'm still curious: is there a plan to set an upper limit on cooldown? Otherwise, setting the value to max uint256 would essentially disable changing the delegate's address though it would still allow its revocation

Author: No, we don't plan to do it. Maybe it could be useful in some edge cases
e.g. if only revocation is required. The owner can't rotate the oracle address, but can disable it.

Reviewer 3: Should we gate in the factory as a policy still?

Author: Since this factory could be used elsewhere across the protocol, I don't think we should limit it with Oracle/Council policies.

For example, I'm not sure we need a cooldown for the depositor bot.

---

#### Part 2: KDF Adoption — Oracle and DSM Integration

This section defines which components are migrated to the delegation model in this release, the concrete changes required for each, and the operator migration process. Adoption is mandatory for Oracle and DSM roles.

##### Scope of Migration in This Release

| Component                              | Integration style                                                                                                                                             |
|----------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Lido Oracle                            | Push (`execute()`)                                                                                                                                            |
| `DepositSecurityModule` contract       | Pull (DSM verifies via a standard ERC-1271 `isValidSignature` check against the guardian's `DelegationContract`)                                              |

> | Council daemon                         | Pull (signs messages, publishes its `DelegationContract` address with each signature, DSM verifies)                                                           |

Reviewer 4: I think the council daemon itself uses only push model (for pause and unvetting) and pull model is used via deposit/pauser/unvet bots for deposits, pause and unvetting

All components in the table are migrated together.

---

###### Oracle Daemon — Push Integration

**Pattern:** Push via `execute()`.

**Authorization model:** `HashConsensus` and the report processors (`AccountingOracle`, `ValidatorsExitBusOracle`, CM's `FeeOracle`, and CSM's `FeeOracle`) check `msg.sender` against the registered oracle committee member set. After migration, `msg.sender` is the `DelegationContract` address, which governance must have registered as the committee member.

**Daemon configuration:** `DELEGATION_CONTRACT_ADDRESS` environment variable. When set, all oracle contract calls are routed through `DelegationModule`, a Web3py extension that wraps `execute()` dispatch. Startup validation checks: (a) `getDelegate() == configuredHotKey`, and (b) `delegationContract` address holds oracle committee membership in `HashConsensus`.

**Transitional behavior:** When `DELEGATION_CONTRACT_ADDRESS` is unset, calls are sent directly from the hot key EOA as before. This path exists for development, for the migration window, and as a long-term fallback.

---

> ###### DSM Contract — Pull Integration

Reviewer 4: Should VERSION be bumped for a contract so that the contract interface can be identified by the council daemon?

Author: There will be no changes in interface itself. So no

Reviewer 4: But we're changing Signature -> GuardianSignature

Author: Agree. Added

**Pattern:** Pull — DSM verifies each signature against the supplied guardian with a standard ERC-1271 `isValidSignature` check over a digest **bound to that guardian**.

> **Authorization model today.** `DepositSecurityModule` does **not** take the guardian address as an input — it *derives* it: every verification path calls `ECDSA.recover(msgHash, sig.r, sig.vs)` and looks the recovered EOA up in the guardian set via `_isGuardian`.

Reviewer 4: It's worth noting that this applies only to deposits and fallbacks for pause and unvetting. Primary way for pause and unvettign is a direct call from the council daemon.

Author: added

**Required change — the guardian address must be supplied with the signature.** A contract guardian cannot be recovered from a signature, so the submitter provides the registered guardian address **explicitly alongside each signature**. Every guardian is a `DelegationContract` — EOA guardians are no longer supported — and DSM verifies the signature **through that contract's** [ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) `isValidSignature(hash, sig)`, requiring the `0x1626ba7e` magic-value return.

```solidity
// Per-signature payload: the registered guardian plus an opaque signature blob.
struct GuardianSignature {
    address guardian;    // registered guardian
    bytes   signature;   // opaque blob, forwarded to the guardian's ERC-1271 isValidSignature
}

function _isValidGuardianSignature(
    address guardian,
    bytes32 msgHash,
    bytes calldata signature
) internal view returns (bool) {
    if (!_isGuardian(guardian)) return false;

>     // Every guardian is a DelegationContract: verify via its ERC-1271

Reviewer 4: I think it would be better to abstract away knowledge of DelegationContract in DSM. From DSM's perspective, it's enough to know that guardians are some contracts that support 1271.

Reviewer 4: We can add a check for support of interface 1271 to the `addGuardian` method

Author: Done

    // isValidSignature, which resolves the effective delegate and fails
    // closed (returns a non-magic value) if there is none.

>     return IERC1271(guardian).isValidSignature(msgHash, signature) == 0x1626ba7e;

Reviewer 4: ```suggestion
    return IERC1271(guardian).isValidSignature(msgHash, signature) == EIP1271_MAGIC_VALUE;
```

Author: done

}
```

**Binding the digest and counting distinct signers.** The legacy EOA scheme got two properties for free that the indirection breaks, and both must be restored explicitly:

1. **The signed digest must be bound to the guardian.** Because the signer (a delegate) is no longer the same address as the principal being counted (the guardian contract), an unbound signature would validate for *every* guardian contract that currently resolves to that delegate. The signed digest therefore includes the guardian address. The existing per-action message prefix already binds the chain id, the DSM address, and the action type — `keccak256(ATTEST_MESSAGE, block.chainid, address(this))` — so adding the guardian completes the binding (guardian + DSM + chain id + action + the existing message fields).

> 2. **Quorum must count distinct guardians and distinct hot keys.** Both dimensions must be unique within a batch. Deduplicating by guardian address alone is insufficient — two different guardian contracts can share one delegate, so a single compromised hot key could still count multiple times toward the 4-of-6 quorum. Deduplicating by delegate alone is also insufficient — it would not stop one guardian from being submitted twice (e.g., if a delegation contract variant supported multiple delegates). DSM therefore rejects both a repeated **guardian address** and a repeated **delegate address** (`getDelegate()`): it requires the batch to be strictly ascending by guardian address (which rejects an unsorted array and any duplicate guardian in a single comparison) and additionally rejects any repeated resolved delegate across the batch:

Reviewer 4: I think it’s sufficient to check only the uniqueness of the guardians’ addresses.

Checking the uniqueness of hot keys adds complexity to the signature quorum check and breaks the abstraction – DSM needs to know that guardians are DelegationContracts and that they have delegates. Although, for DSM, it’s enough to simply know that guardians support 1271.

The check itself only protects against a very unusual scenario where multiple guardians delegate their authority to the same EOA. Such a scenario is impossible in a real-world setup where guardians are distinct entities. 

An operational error does not lead to any consequences (if Guardian A has delegated to Guardian B’s hot key, Guardian B will not sign on behalf of Guardian A, since its setup is not configured for this – `DELEGATION_CONTRACT_ADDRESS` points to its own DelegationContract address). And such a configuration mistake is enough to be mitigated by alerting.

A deliberate attack in this specific form doesn’t make much sense. Since having full access to the keys and delegation allows one to perform any signatures for any operations

**Required change to the council daemon:** Sign with the delegate private key, set `DELEGATION_CONTRACT_ADDRESS`, and publish that `DelegationContract` address together with each signature.

**Required change to the depositor bot:** The bot relays the council-provided `(DelegationContract address, signature)` pairs to DSM as-is. Two behavioral changes are required:

1. **Filtering (on receival):** accept an incoming message only if its signer is the **current delegate** of the referenced guardian's `DelegationContract` (resolved via `getDelegate()`), instead of matching against a delegation-contract address. Messages signed by a rotated-out key are dropped.
2. **Ordering (on submission):** when assembling `depositBufferedEther`, sort the signature array by **guardian address**, matching DSM's strictly-ascending-by-guardian check.

---

###### Validator Ejector (Node-Operator Side) — Off-Chain Update

The **validator ejector** is the off-chain daemon run by each node operator that watches `ValidatorsExitBusOracle` (VEBO) — the oracle that publishes which validators must exit — and broadcasts the corresponding signed voluntary-exit messages to the consensus layer.

> To accept those messages the ejector keeps an **allowlist** of trusted oracle signing addresses. Oracle messages are always signed by the **delegate EOA**, never by the `DelegationContract`, so the allowlist must contain the **delegate EOA** and be refreshed to the new EOA whenever the delegate is rotated.

Reviewer 4: We should set up some kind of automation for address rotation, since we won’t be controlling this process anymore. For example, we could send an alert to the Lido Validators channel for each rotation, including the current list of EOA addresses and a request to update those addresses in their envs.

Reviewer 4: We could come up with some kind of automation for this. For example, we could sign a list of addresses with our own private key, let's say in CI/CD, and provide the node operators with just a single public key instead of the entire list and automate the list updating. It’s less secure (because it's based just on one private key), but just as an example.

But since we’ll be getting rid of Ejector altogether soon, there’s no need to worry too much about this. It’s just a temporary measure.

Author: Yeah, I think the alerting system should be enough for NOMs to announce until the ejector is deprecated.

| `DelegationContract` (one per bot) | each operator, via the factory | One instance per Oracle/Council bot; deployed during the operator migration (Part 2), not by a single governance action. |

**Redeployed at a new address (standalone, non-upgradeable) — references must be repointed by governance:**

| Contract                | Repointing required                 |
|-------------------------|-------------------------------------|
| `DepositSecurityModule` | Update the `LidoLocator` reference. |

## Links

- Delegation contracts repository (`lidofinance/delegation-execution-authority`): https://github.com/lidofinance/delegation-execution-authority
- Oracle daemon (`lidofinance/lido-oracle`): https://github.com/lidofinance/lido-oracle
- Core repository (`lidofinance/core`, contains `DepositSecurityModule`): https://github.com/lidofinance/core

## Copyright

```my python
                           (o)(o)
                          /     \
                         /       |
                        /   \ .. |
          ________     /    /\__/
  _      /        \   /    /   \\
 / \    /  ____    \_/    /
//\ \  /  /    \         /
V  \ \/  /      \       /
    \___/        \_____/

> ```

Reviewer 4: Can it be hidden with html comments?) And replaced with canonical copyright

### A standard primitive for permissioned key handling

Today each role handles its signing keys in its own one-off way, and rotating any of them runs through the same heavyweight governance path. Without a shared primitive, every new role re-invents key custody and rotation, and monitoring must learn each role's own conventions — multiplying operational risk. Standardizing delegation as a single, audited mechanism lets any role inherit the same rotation, revocation, and termination semantics, and lets monitoring treat every delegated key uniformly.

KDA improvements unlock:

- **Routine rotation as hygiene.** Periodic hot-key rotation limits the value of any single leaked key and shrinks the window in which an undetected compromise stays useful — a standard operational practice that the governance-gated path makes impractical today.

> - **Owners’ setup** Owners increasingly prefer a multisig (or hard-wallet) over a bare EOA for stronger custody.

Reviewer: ```suggestion
- **Owners’ setup.** Owners increasingly prefer a multisig (or hard-wallet) over a bare EOA for stronger custody.
```

    ///         Only callable by owner. The new delegatee becomes effective only
    ///         after the contract's cooldown (`getCooldown()` seconds, or
    ///         immediately if cooldown is 0). The currently effective delegatee
    ///         (if any) stays effective throughout the cooldown and is dropped
    ///         only when the new one activates — so a planned rotation is
    ///         seamless, with no interval lacking an active delegatee.
    ///         Reassigning before the cooldown elapses replaces the pending
    ///         delegatee and restarts the cooldown; the current one stays
    ///         effective throughout. To drop a (e.g. compromised) delegatee
    ///         immediately, use revokeDelegatee().
    ///         Reverts if delegatee == owner.
    ///         Reverts if the contract is terminated.
    /// @param delegatee Address of the incoming delegatee.
    function assignDelegatee(address delegatee) external;

    /// @notice Remove the current and pending delegatee. 
    ///         Only callable by owner.
    ///         Reverts if the contract is terminated.
    function revokeDelegatee() external;

>     /// @notice Terminate the contract, permanently disabling execute().

Reviewer: ```suggestion
    /// @notice Terminate the contract, permanently disabling execute(), signature verification via isValidSignature(), and further delegate reassignment via assignDelegate.
```

    ///         Only callable by owner.
    ///         Also clears the active delegatee (as revokeDelegatee), so
    ///         getDelegatee() returns address(0) after termination.
    ///         Intended for emergency use when the owner is suspected compromised.
    ///         Termination is irreversible.
    ///         Reverts if the contract is already terminated.
    function terminate() external;

    // --- Push integration ---

    /// @notice Execute an arbitrary non-delegate call on behalf of this contract.
    ///         Only callable by the current delegatee.
    ///         Reverts if the contract is terminated.
    ///         Reverts if the target call reverts.
    ///         Forwards msg.value to the target to support payable targets.
    ///         After the target call returns, any residual ETH balance on

>     ///         this contract is swept back to the delegatee (msg.sender).

Reviewer: ```suggestion
    ///         After the target call returns, any residual ETH provided as a payment 
    ///         is swept back to the delegatee (msg.sender).
```

    /// @param target  Address to call.
    /// @param data    Call data.
    /// @return result Return data from the call.
    function execute(address target, bytes calldata data)
        external
        payable
        returns (bytes memory result);

    // --- Asset recovery ---
    //
    // Rescue assets that were sent to this contract (the delegatee can already
    // move them via execute(); these give the cold owner an independent path,
    // usable even after the delegatee is revoked or the contract is terminated).

>     // Only callable by owner. `recipient` must be non-zero.

Reviewer: delegate should not be able to touch funds in any case

    /// @notice Recover ETH held by this contract to `recipient`.
    function recoverEther(address recipient, uint256 amount) external;

    /// @notice Recover an ERC-20 balance held by this contract to `recipient`.
    function recoverERC20(address token, address recipient, uint256 amount) external;

    /// @notice Recover an ERC-721 token held by this contract to `recipient`.
    function recoverERC721(address token, uint256 tokenId, address recipient) external;

    /// @notice Recover an ERC-1155 balance held by this contract to `recipient`.

>     function recoverERC1155(address token, uint256 tokenId, address recipient, uint256 amount) external;

Reviewer: Move it to the bottom of the interface

    ///         `IDelegationContract` interface ids.
    function supportsInterface(bytes4 interfaceId) external view returns (bool);

    // --- Views ---

    /// @notice The contract's owner. Provided as the
    ///         read-only ERC-5313 ownership view so explorers, multisig UIs,
    ///         and generic tooling recognize the controlling party.
    function owner() external view returns (address);

    /// @notice Returns the currently *effective* delegatee, or address(0) if
    ///         none. After assignDelegatee(), the previously effective delegatee
    ///         remains the effective one until the new delegatee's cooldown
    ///         elapses; only then does this return the new delegatee. Returns
    ///         address(0) when there is no current delegatee (never assigned, or
    ///         revoked) and once the contract is terminated.
    function getDelegatee() external view returns (address);

    /// @notice Returns the pending (not-yet-effective) delegatee and the
    ///         timestamp at which it becomes effective, or (address(0), 0) if
    ///         no assignment is in cooldown.

>     function getPendingDelegatee() external view returns (address delegatee, uint256 activeFrom);

Reviewer: We need to think about cleaning the pending delegate after it becomes active.

    /// @notice Cooldown in seconds between assigning a delegatee and it
    ///         becoming effective. Set in the constructor (Solidity immutable)
    ///         and unchangeable thereafter.
    function getCooldown() external view returns (uint256);

    /// @notice Returns true if the contract has been terminated.
    function isTerminated() external view returns (bool);

    // --- Events ---

>     event DelegateeNominated(address indexed newDelegatee, uint256 activeFrom);

Reviewer: I suggest also having DelegateChanged(address indexed newDelegatee) to clearly capture the moment the new delegate becomes active. I think we can have a permissionless method that will apply pending delegate when the time has come. It is needed to ensure that isValidSignature remains a view function with a deterministic output (time independent).

Reviewer 2: Im not sure, a few points:
- DelegateeNominated(newDelegatee, activeFrom) carries activeFrom, and cooldown is known, so anyone can compute the effective delegatee at any block without an extra steps.
-  Lazy activation makes the new key effective exactly at activeFrom with no coordination, which matches KDF's goal of rotation being a cheap, simple action. The additional action will not be as convenient for holders and I don’t want them to forget about it.

At first glance, it seems to add more randomness. Activation will be is no longer calculated but relies on human factor or bot timings.

        │                                  │   principal (e.g. guardian)  │
```

**Security assumptions:**
- The integrator must treat the `DelegationContract` address as the authorized principal: it checks that the supplied `DelegationContract` is a registered entity **and** that `isValidSignature` returns the magic value (`0x1626ba7e`) for a valid signature.
- The signing key must match the effective delegatee at the time of on-chain verification. A key rotated (or the contract terminated) after signing but before verification causes `isValidSignature` to reject — failing closed.
- The fail-closed-on-zero-delegatee guard lives **inside** `isValidSignature` (it returns the failure value rather than verifying against `address(0)`), so integrators cannot accidentally validate a malformed signature against an unassigned delegatee.

Reference implementation of the contract side:

```solidity
function isValidSignature(bytes32 hash, bytes calldata signature)
    external
    view
    returns (bytes4)
{
    address delegatee = getDelegatee();   // effective delegatee, or address(0)
    // Fail closed when there is no effective delegatee (never assigned/revoked/terminated).

>     if (delegatee != address(0) && SignatureChecker.isValidSignatureNow(delegatee, hash, signature)) {

Reviewer: Add a note that SignatureChecker is - https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5c70e564370f70647ab55b79508904a039a123f1/contracts/utils/cryptography/SignatureChecker.sol#L7

>         return IERC1271.isValidSignature.selector;   // 0x1626ba7e

Reviewer: I can't find this item in https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/interfaces/IERC1271.sol so pls just define 2 constants as it is done in the actual code now

Reviewer 2: Done

- A new `DelegationContract` must be deployed and the role must be reassigned to the new address.

Termination is intended for scenarios where the owner believes the owner's own address has been compromised and wishes to ensure no further actions can be taken under its authority.

##### Hot-Key Rotation

For a **planned** rotation the owner calls `assignDelegatee(newHotKey)`. The previous key keeps operating until the new key becomes effective after the cooldown, so the rotation is seamless.

The specific cadence for routine rotation is not mandated here; recommended practices will be maintained on the Lido research forum and may be revised. Contract-level requirements are:

- On any suspected or confirmed hot-key compromise, the owner must call `revokeDelegatee()` to drop the key **immediately**. 
- A replacement is then assigned with `assignDelegatee(newHotKey)` and becomes effective after its cooldown.
- After assignment, the operator updates the daemon configuration.

##### Asset Recovery

> The `DelegationContract` is not meant to custody assets, but ETH or tokens can still arrive at its address — change from a payable `execute()` call, an accidental transfer, or an airdrop. To avoid stranding them, the owner can rescue them with `recoverEther` / `recoverERC20` / `recoverERC721` / `recoverERC1155`, each sending to an owner-specified, non-zero `recipient`.

Reviewer: Change should be handled within execute. We should have receive() method that reverts. But it will still leave an opportunity to do self-distract and send ETH. For non ETH assets, there is no way to stop someone from making a transfer.

> Recovery is **owner-gated**, not delegatee-gated, for two reasons. First, the delegatee can *already* move any asset out via `execute(token, transferCalldata)`, so dedicated recovery methods add no new capability for the hot key — they exist to give the **cold owner an independent path** that works even after the delegatee is revoked or the contract is terminated (when `execute()` is disabled). Second, keeping asset movement on the cold key matches the rest of the model: the hot delegatee acts on the protocol, the owner controls value-bearing escape hatches. These methods do not require the contract to implement any token-receiver hook (ERC-721/1155 `onReceived`), and adding them does not make the contract a custodian — they only rescue what happens to land there.

Reviewer: Hmmmm. So maybe I was wrong, and there is no need for these methods since delegate can indeed transfer any token. Then I suggest we do not add them. The only one that we can add is receive Ether (so that we will not hide it within execute refund)
